File: cpp/graphAlgorithms/spath/spt_d.cpp
/** @file spt_d.cpp
 * 
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "stdinc.h"
#include "Heap_d.h"
#include "Graph_wd.h"

namespace grafalgo {

/** Find a shortest path tree using Dijkstra's algorithm.
 *  @param g is a directed graph with edge lengths
 *  @param s is the source vertex for the shortest path tree computation;
 *  if s=0, paths are computed from an imaginary extra vertex with a
 *  zero length edge to every other vertex
 *  @param pEdge is an array of parent pointers; on return pEdge[u] is the
 *  number of the edge connecting u to its parent in the shortest path tree
 *  @param d is array of distances; on return d[u] is the shortest path
 *  distance from s to u
 *  @return true on success, false if not all vertices can be reached
 */
bool spt_d(Graph_wd& g, vertex s, edge* pEdge, edgeLength* d) {
	Heap_d<int> nheap(g.n(),4);

	for (vertex v = 1; v <= g.n(); v++) { pEdge[v] = 0; d[v] = INT_MAX; }
	d[s] = 0; nheap.insert(s,0);
	int cnt = 0;
	while (!nheap.empty()) {
		vertex v = nheap.deletemin(); cnt++;
		for (edge e = g.firstOut(v); e != 0; e = g.nextOut(v,e)) {
			vertex w = g.head(e);
			if (d[w] > d[v] + g.length(e)) {
				d[w] = d[v] + g.length(e); pEdge[w] = e;
				if (nheap.member(w)) nheap.changekey(w, d[w]);
				else nheap.insert(w, d[w]);
			}
		}
	}
	return cnt == g.n();
}

} // ends namespace

File: js/graphAlgorithms/spath/sptD.mjs
/** @file sptD.mjs
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import { assert, EnableAssert as ea } from '../../common/Assert.mjs';
import ArrayHeap from '../../dataStructures/heaps/ArrayHeap.mjs';

/** Compute shortest path tree of a graph using Dijkstra's algorithm.
 *  @param g is a digraph with edge lengths
 *  @param s is a "source vertex" in g
 *  @param trace controls output of information about the internal  
 *  state of the computation; larger values produce more information
 *  @return a tuple [link, dist, ts, stats] where link[u] is the edge from
 *  the parent of u to u in the spt rooted at s (or 0 if u unreachable);
 *  dist[u] is the shortest path distance from vertex s to vertex u
 *  (or infinity if u unreachable), ts is a trace string and stats is a
 *  statistics object; if g has negative edges, returns []
 */
export default function sptD(g, s, trace=0) {
	let link = new Int32Array(g.n+1); let ts = '';
	let dist = new Array(g.n+1).fill(Infinity);
	let border = new ArrayHeap(g.n, 2+Math.floor(g.m/g.n));
	if (trace) ts += g.toString(1);

	dist[s] = 0; border.insert(s, 0);
	if (trace) ts += 'initial heap: ' + border + '\n\n' +
					 'selected vertex, edge to parent, ' + 
					 `distance from ${g.x2s(s)}, heap contents\n`;
	while (!border.empty()) {
		let u = border.deletemin();
		for (let e = g.firstOutof(u); e; e = g.nextOutof(u, e)) {
			if (g.length(e) < 0) return [];
			let v = g.head(e);
			if (dist[v] > dist[u] + g.length(e)) {
				dist[v] = dist[u] + g.length(e); link[v] = e;
				if (border.contains(v)) border.changekey(v, dist[v]);
				else border.insert(v, dist[v]);
			}
		}
		if (trace) {
			ts += g.x2s(u) + ' ' + (link[u] > 0 ? g.e2s(link[u]) : '-') +
				   ' ' + dist[u] + ' ' + border + '\n';
		}
	}
	return [link, dist, ts,  border.getStats()];
}
